### PR TITLE
Terrain doc

### PR DESCRIPTION
**Pull request by AlemSnyder**
Adds documentation to six files. These are the four in Terrain generation, and two others.

### CODE REVIEW

File: src/Terrain/TerrainGeneration/material.hpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ * The material determines if the tile is solid, and the potential color. Other
+ * data will be added like how cretin materials respond to weather...
+ */
+struct Material{
+    // vector of <name hex color> for possible colors
+    std::vector<std::pair<const char *, uint32_t>> color;
+    uint8_t speed_multiplier = 1; // speed on this material compared to base
+    bool solid = false; // Is the material solid?
+    uint8_t element_id = 0; // The ID of the material (Air is 0)
+    const char * element = "Air"; // The material name (why is this not name?)
```
Comment (MrAwesomeRocks) [inline src/Terrain/TerrainGeneration/material.hpp:62]:
```suggestion
    const std::string element = "Air"; // The material name (why is this not name?)
```

File: src/Terrain/TerrainGeneration/material.hpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ * @details World class should have a map of materials organized by Material
+ * Id. Each tile has a pointer to one of those materials (though this should
+ * be an uint8_t as there are less than 256 materials).
+ * 
+ * The material determines if the tile is solid, and the potential color. Other
+ * data will be added like how cretin materials respond to weather...
+ */
+struct Material{
+    // vector of <name hex color> for possible colors
+    std::vector<std::pair<const char *, uint32_t>> color;
```
Comment (MrAwesomeRocks) [inline src/Terrain/TerrainGeneration/material.hpp:58]:
```suggestion
    std::vector<std::pair<const std::string, uint32_t>> color;
```

File: src/Terrain/chunk.cpp [hunk 3]
```diff
@@ -1,25 +1,24 @@
...
-Chunk::Chunk(int bx, int by, int bz, Terrain * ter_){
-    ter = ter_;
-    Cx=bx;
-    Cy=by;
-    Cz=bz;
+Chunk::Chunk(int bx, int by, int bz, Terrain * ter){
+    ter_ = ter;
+    Cx_=bx;
+    Cy_=by;
+    Cz_=bz;
```
Comment (MrAwesomeRocks) [inline src/Terrain/chunk.cpp:10]: b/C???

File: src/Terrain/terrain.cpp [hunk 4]
```diff
@@ -345,7 +345,15 @@ void Terrain::stamp_tile_region(int x_start, int y_start, int z_start, int x_end
...
-    stamp_tile_region(tStamp.x_start + x*Area_size + Area_size/2, tStamp.y_start + y*Area_size + Area_size/2 , tStamp.z_start, tStamp.x_end + x*Area_size + Area_size/2, tStamp.y_end + y*Area_size + Area_size/2, tStamp.z_end, tStamp.mat, tStamp.elements_can_stamp, tStamp.color_id);
+    stamp_tile_region(tStamp.x_start + x*Area_size + Area_size/2,
+                      tStamp.y_start + y*Area_size + Area_size/2,
+                      tStamp.z_start,
+                      tStamp.x_end + x*Area_size + Area_size/2,
+                      tStamp.y_end + y*Area_size + Area_size/2,
+                      tStamp.z_end,
+                      tStamp.mat,
+                      tStamp.elements_can_stamp,
+                      tStamp.color_id);
```
Comment (MrAwesomeRocks) [inline src/Terrain/terrain.cpp:356]: jeez why did you expand the vector

**Review by MrAwesomeRocks — CHANGES_REQUESTED** [review]
Make the `Terrain` folder lowercase

File: src/Terrain/terrain.cpp (hunk 4 above)
Comment (AlemSnyder) [reply src/Terrain/terrain.cpp:356]: Because it needs to be unpacked. There is not much I can do for this. If it ant broke don't fix it.

**Comment by MrAwesomeRocks** [comment]
You need to use enums in terrain.cpp, not just `type == 8`, etc.

File: src/Terrain/terrain.cpp [hunk 5]
```diff
@@ -595,15 +593,15 @@ const OnePath Terrain::get_path_type(int xs, int ys, int zs, int xf, int yf, int
     int dz = 3;
     int dxy = 1;
 
-    int8_t type = abs(xs - xf) + abs(ys - yf) + 4 * abs(zs - zf);
+    OnePath type(((abs(xs - xf) + abs(ys - yf)) << (1 + 2 * abs(zs - zf))) + (abs(zs - zf) << 1));
```
Comment (MrAwesomeRocks) [inline src/Terrain/terrain.cpp:596]: Ok you need to comment this.

And not put it in the initializer

**Review by MrAwesomeRocks — CHANGES_REQUESTED** [review]
Almost done

File: src/Terrain/terrain.cpp (hunk 5 above)
Comment (AlemSnyder) [reply src/Terrain/terrain.cpp:596]: lol

File: CMakeLists.txt [hunk 6]
```diff
@@ -1,6 +1,9 @@
 cmake_minimum_required(VERSION 3.10)
 #add_compile_options(-municode)
 
+# set policies
```
Comment (MrAwesomeRocks) [inline CMakeLists.txt:4]:
```suggestion
# prefer newer OpenGL
# https://cmake.org/cmake/help/latest/policy/CMP0072.html
```

File: CMakeLists.txt [hunk 7]
```diff
@@ -10,7 +13,7 @@ project(Fun_Game)
 
 find_package(OpenGL REQUIRED)
 find_package(GLEW REQUIRED)
-find_package(glfw3 3.3 REQUIRED)
+find_package(glfw3 REQUIRED)
```
Comment (MrAwesomeRocks) [inline CMakeLists.txt:16]: Version pinning is good, why not use it?

File: CMakeLists.txt [hunk 8]
```diff
@@ -20,10 +23,10 @@ file(GLOB SOURCES3 "src/**/**/*.cpp")
 
 add_executable(Fun_Game ${SOURCES1} ${SOURCES2} ${SOURCES3})
 
-target_link_libraries(Fun_Game glfw3)
+target_link_libraries(Fun_Game glfw)
```
Comment (MrAwesomeRocks) [inline CMakeLists.txt:26]: Again, version pinning

File: CMakeLists.txt [hunk 9]
```diff
@@ -20,10 +23,10 @@ file(GLOB SOURCES3 "src/**/**/*.cpp")
...
 target_link_libraries(Fun_Game OpenGL::GL)
 target_link_libraries(Fun_Game GLEW::GLEW)
-target_link_libraries(Fun_Game glm)
+#target_link_libraries(Fun_Game glm)
```
Comment (MrAwesomeRocks) [inline CMakeLists.txt:29]: Helps prevent compile errors (gives better error message for missing library), don't remove.

File: CMakeLists.txt [hunk 10]
```diff
@@ -42,6 +45,8 @@ endif()
 #  "${CMAKE_CXX_FLAGS} -municode"
 #)
 
+set(CMAKE_BUILD_TYPE Debug)
+
```
Comment (MrAwesomeRocks) [inline CMakeLists.txt:49]: NO THIS IS SET THOUGH THE BUILD COMMAND DON'T DO THIS

File: src/Terrain/terrain.cpp [hunk 11]
```diff
@@ -164,8 +164,8 @@ void Terrain::init_area(int area_x, int area_y, LandGenerator gen){
 
 void Terrain::init_chunks(){
     for (int xyz=0; xyz<((X_MAX-1)/Chunk::size+1)*((Y_MAX-1)/Chunk::size+1)*((Z_MAX-1)/Chunk::size+1); xyz+=1){
-        auto [x,y,z] = sop(xyz, ((X_MAX-1)/Chunk::size+1),((Y_MAX-1)/Chunk::size+1),((Z_MAX-1)/Chunk::size+1));
-        chunks.push_back(Chunk(x,y,z, this));
+        auto xyz_sop = sop(xyz, ((X_MAX-1)/Chunk::size+1),((Y_MAX-1)/Chunk::size+1),((Z_MAX-1)/Chunk::size+1));
+        chunks.push_back(Chunk(xyz_sop[0],xyz_sop[1],xyz_sop[2], this));
```
Comment (MrAwesomeRocks) [inline src/Terrain/terrain.cpp:168]: What's wrong with destructuring? Did you set your C++ version?

https://en.cppreference.com/w/cpp/language/structured_binding

File: src/Terrain/terrain.cpp [hunk 12]
```diff
@@ -448,8 +448,8 @@ std::vector<int> Terrain::generate_macro_map(unsigned int size_x, unsigned int s
     NoiseGenerator ng = NoiseGenerator(numOctaves, persistance, 3);
 
     for (unsigned int i = 0; i < out.size(); i++){
-        auto [x,y,z] = sop(i,size_x,size_y,1);
-        auto p = ng.getValueNoise((double)x*spacing,(double)y*spacing);
+        auto xyz = sop(i,size_x,size_y,1);
```
Comment (MrAwesomeRocks) [inline src/Terrain/terrain.cpp:451]: What's wrong with destructuring? Did you set your C++ version?

https://en.cppreference.com/w/cpp/language/structured_binding

File: src/Terrain/terrain.cpp [hunk 13]
```diff
@@ -473,25 +473,25 @@ float Terrain::get_H_cost(std::array<float, 3> xyz1, std::array<float, 3> xyz2)
     double D1 = 1.0;
     double D2 = 1.414;
     double D3 = 1.0;
-    auto [x1, y1, z1] = xyz1;
-    auto [x2, y2, z2] = xyz2;
+    //auto [x1, y1, z1] = xyz1;
+    //auto [x2, y2, z2] = xyz2;
```
Comment (MrAwesomeRocks) [inline src/Terrain/terrain.cpp:477]: What's wrong with destructuring? Did you set your C++ version?

https://en.cppreference.com/w/cpp/language/structured_binding

File: src/Terrain/terrain.cpp [hunk 14]
```diff
@@ -473,25 +473,25 @@ float Terrain::get_H_cost(std::array<float, 3> xyz1, std::array<float, 3> xyz2)
...
 
-    float DX = abs(x1 - x2);
-    float DY = abs(y1 - y2);
-    float DZ = abs(z1 - z2);
+    float DX = abs(xyz1[0] - xyz2[0]);
+    float DY = abs(xyz1[1] - xyz2[1]);
+    float DZ = abs(xyz1[2] - xyz2[2]);
```
Comment (MrAwesomeRocks) [inline src/Terrain/terrain.cpp:481]: Old names were better

File: src/Terrain/terrain.cpp [hunk 15]
```diff
@@ -473,25 +473,25 @@ float Terrain::get_H_cost(std::array<float, 3> xyz1, std::array<float, 3> xyz2)
...
```
Comment (MrAwesomeRocks) [inline src/Terrain/terrain.cpp:494]: Old names were better

File: src/Terrain/terrain.cpp [hunk 16]
```diff
@@ -473,25 +473,25 @@ float Terrain::get_H_cost(std::array<float, 3> xyz1, std::array<float, 3> xyz2)
...
     return (DZ * D3 + abs(DX - DY) * D1 + D2 * std::min(DX, DY));
 }
 float Terrain::get_H_cost(std::array<int, 3> xyz1, std::array<int, 3> xyz2) {
     double D1 = 1.0;
     double D2 = 1.414;
     double D3 = 1.0;
-    auto [x1, y1, z1] = xyz1;
-    auto [x2, y2, z2] = xyz2;
+    //auto [x1, y1, z1] = xyz1;
+    //auto [x2, y2, z2] = xyz2;
```
Comment (MrAwesomeRocks) [inline src/Terrain/terrain.cpp:490]: What's wrong with destructuring? Did you set your C++ version?

https://en.cppreference.com/w/cpp/language/structured_binding

File: src/Terrain/terrain.cpp [hunk 17]
```diff
@@ -507,13 +507,19 @@ void Terrain::add_all_adjacent(int xyz) {
...
-        auto [x_, y_, z_] = sop(xyz);
-        auto [xs, ys, zs] = sop(xyz_, 3, 3, 3);
+        auto xyz_center = sop(xyz);
+       	int x_c = xyz_center[0];
+        int y_c = xyz_center[0];
+        int z_c = xyz_center[2];
+        auto xyz_direction = sop(xyz_, 3, 3, 3);
+        int x_d = xyz_direction[0]; // direction to final tile. can be +1, -1, or 0
+        int y_d = xyz_direction[0];
+        int z_d = xyz_direction[2];
```
Comment (MrAwesomeRocks) [inline src/Terrain/terrain.cpp:517]: What's wrong with destructuring? Did you set your C++ version?

https://en.cppreference.com/w/cpp/language/structured_binding

**Review by MrAwesomeRocks — CHANGES_REQUESTED** [review]
Uhh have some questions

File: CMakeLists.txt (hunk 10 above)
Comment (AlemSnyder) [reply CMakeLists.txt:50]: lol I forgot I did this. oops

File: CMakeLists.txt (hunk 9 above)
Comment (AlemSnyder) [reply CMakeLists.txt:30]: glm isn't a library its a header file (I think). For some reason this breaks linux.

File: CMakeLists.txt (hunk 8 above)
Comment (AlemSnyder) [reply CMakeLists.txt:27]: this breaks linux

File: CMakeLists.txt (hunk 9 above)
Comment (MrAwesomeRocks) [reply CMakeLists.txt:30]: The entire Standard Template *Library* is made up of headers. Header-only libraries are used with templates.

https://stackoverflow.com/questions/30055777/what-is-a-header-only-library

File: CMakeLists.txt (hunk 8 above)
Comment (MrAwesomeRocks) [reply CMakeLists.txt:27]: The change may break windows though

Comment (AlemSnyder) [reply CMakeLists.txt]: it did not

Comment (MrAwesomeRocks) [reply CMakeLists.txt]: Ok

File: CMakeLists.txt [hunk 18]
```diff
@@ -1,3 +1,7 @@
+# specify the C++ standard
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD_REQUIRED True)
+
```
Comment (MrAwesomeRocks) [inline CMakeLists.txt:4]: `cmake_minimum_required` is always first

File: CMakeLists.txt [hunk 19]
```diff
@@ -28,10 +27,12 @@ file(GLOB SOURCES3 "src/**/**/*.cpp")
 
 add_executable(Fun_Game ${SOURCES1} ${SOURCES2} ${SOURCES3})
 
+#target_include_directories(Fun_Game ${GLM_INCLUDE_DIRS})
+
```
Comment (MrAwesomeRocks) [inline CMakeLists.txt:31]: Delete this comment

**Review by MrAwesomeRocks — APPROVED** [review]
LGTM

**Pull request merged**